internal/storage/asset_store.go

```
@@ -0,0 +1,196 @@
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+
+	"platform-go-challenge/internal/common"
+	"platform-go-challenge/internal/models"
+)
+
+// FavoritesStore to manage user favorite assets
+// I have chosen to use a map of maps to represent user favorites for two main reasons:
+//  1. Avoid circular dependencies: By defining FavoritesStore in the storage package, we avoid circular import issues between models and storage packages. This keeps the codebase modular and easier to maintain.
+//  2. In a relational database, favorites would typically be represented as a join table between users and assets. Using a map of maps simulates this relationship effectively,
+//     allowing quick lookups and management of user favorites without complicating the User model itself.
+type FavoritesStore struct {
+	favorites map[uuid.UUID]map[string]models.Asset // Map of UserID to map of AssetID to Asset
+	userLocks map[uuid.UUID]*sync.RWMutex           // Per-user locks for concurrency safety
+	mutex     sync.RWMutex                          // Global lock for the FavoritesStore
+}
+
+func NewFavoritesStore() *FavoritesStore {
+	return &FavoritesStore{
+		favorites: make(map[uuid.UUID]map[string]models.Asset),
+		userLocks: make(map[uuid.UUID]*sync.RWMutex),
+	}
+}
+
+// ---------- Favortites management methods ----------
+
+func (fs *FavoritesStore) AddFavorite(userID uuid.UUID, favorite models.Asset) error {
+	fs.mutex.Lock()
+
+	if _, exists := fs.favorites[userID]; !exists {
+		fs.userLocks[userID] = &sync.RWMutex{}
+		fs.favorites[userID] = make(map[string]models.Asset)
+	}
+	userLock := fs.userLocks[userID]
+	fs.mutex.Unlock()
+
+	// Lock per user
+	userLock.Lock()
+	defer userLock.Unlock()
+
+	if _, exists := fs.favorites[userID][favorite.GetID()]; exists {
+		return errors.New("asset already in favorites")
```

This error here will lead to a 500 error response to the user. Is there any other status code that could be used for this kind of error?

---

Yes, in retrospect I would return a 409 error response; the client is trying to create a resource that already exists, which conflicts with the current state.